lib/array-based/writable.js

```diff
@@ -0,0 +1,396 @@
+    endWritable(this, state, callback);
+  }
+
+  return this;
+};
+
+Writable.prototype.cork = function() {
+  this._writableState.corked++;
+};
+
+Writable.prototype.uncork = function() {
+  const state = this._writableState;
+
+  if (state.corked) {
+    state.corked--;
+
+    if (
+      !state.writing &&
+      !state.corked &&
+      !state.bufferProcessing &&
+      state.bufferedArray &&
+      Array.isArray(state.bufferedArray) &&
+      state.bufferedSize
+    ) {
+      clearBuffer(this, state);
+    }
+  }
+};
+
+Writable.prototype.pipe = function() {
+  this.emit('error', new Error('Cannot pipe, not readable'));
+};
+
+// internal functions
+
+const writeAfterEnd = (stream, callback) => {
+  const error = new Error('write after end');
+
+  stream.emit('error', error);
+  process.nextTick(callback, error);
+};
+
+function isChunkValid(stream, state, chunk, callback) {
+  let error;
+
+  if (chunk === null) {
+    error = new Error('May not write null values to stream');
+  } else if (typeof chunk !== 'string' && !state.objectMode) {
+    error = new Error('Chunk must be an instance of Buffer or string');
+  }
+
+  if (error) {
+    stream.emit('error', error);
+    process.nextTick(callback, error);
+    return false;
+  }
+
+  return true;
+};
+
+const writeOrBuffer = (stream, state, isBuffer, chunk, encoding, callback) => {
+  if (!isBuffer) {
+    const newChunk = decodeChunk(state, chunk, encoding);
+
+    if (chunk !== newChunk) {
+      isBuffer = true;
+      encoding = 'buffer';
+      chunk = newChunk;
+    }
+  }
+
+  const length = state.objectMode ? 1 : chunk.length;
+  state.length += length;
+
+  let ret = state.length < state.highWaterMark;
+
+  if (!ret) {
+    state.needDrain = true;
+  }
+
+  if (state.writing || state.corked) {
+    if (state.bufferedArray && Array.isArray(state.bufferedArray) && state.bufferedSize) {
+      state.bufferedArray.push({
+        chunk,
+        encoding,
+        isBuffer,
+        callback,
+      });
+    } else {
+      state.bufferedArray = [{
+        chunk,
+        encoding,
+        isBuffer,
+        callback,
+      }];
+    }
+    state.bufferedSize += 1;
+  } else {
+    doWrite(stream, state, false, length, chunk, encoding, callback);
+  }
+
+  return ret;
+};
+
+const doWrite = (stream, state, writev, length, chunk, encoding, callback) => {
+  state.writelen = length;
+  state.writecb = callback;
+  state.writing = true;
+  state.sync = true;
+
+  if (state.destroyed) {
+    state.onwrite(new Error('Cannot call write after a stream was destroyed'));
+  } else if (writev) {
+    stream._writev(chunk, state.onwrite);
+  } else {
+    stream._write(chunk, encoding, state.onwrite);
+  }
+
+  state.sync = false;
+};
+
+const onwrite = (stream, error) => {
+  const state = stream._writableState;
+  const sync = state.sync;
+  const callback = state.writecb;
+
+  if (typeof callback !== 'function') {
+    throw new Error('Callback called multiple times');
```

Comment:
@tshemsedinov Maybe we should not throw mistakes like this. For example, we can take all the error messages in constants, or at least create variables with them somewhere, and then turn to them? Or is it normal?

Reply:
For this certain purpose we have `callback = common.once(callback)` in https://github.com/metarhia/common I do not think we should throw in such cases